### PR TITLE
Percussion panel - layout and interaction basics

### PR DESCRIPTION
**Pull request by mathesoncalum**
Some things to note: 

- There is no interaction with the score yet (hence why an example/placeholder kit is used). This PR focuses purely on the panel layout and UX basics.

-   The “toolbar” (Write, Preview, etc.) is placed within the percussion panel area in this PR. This is expected as we first need to merge [#22050](https://github.com/musescore/MuseScore/pull/22050) to place these controls in the `DockFrame`.

To summarise the interactions/responsibilities of the new components: `PercussionPanelPadModel` will maintain all data relevant to single pad (currently this only consists of the instrument name), `PercussionPanelPadListModel` will handle the initialisation of pads in the grid (and any layout editing), and the `PercussionPanelModel` handles panel modes (and eventually score interactions).

### CODE REVIEW

File: src/notation/view/percussionpanel/percussionpanelpadlistmodel.cpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "percussionpanelpadlistmodel.h"
+
+PercussionPanelPadListModel::PercussionPanelPadListModel(QObject* parent)
+    : QAbstractListModel{parent}
```
Comment (RomanPudashkin) [inline src/notation/view/percussionpanel/percussionpanelpadlistmodel.cpp:26]: We usually use a different style: QAbstractListModel(parent)

File: src/notation/view/percussionpanel/percussionpanelpadlistmodel.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    Q_OBJECT
+
+    Q_PROPERTY(int numColumns READ numColumns CONSTANT)
+public:
+    explicit PercussionPanelPadListModel(QObject* parent = nullptr);
+
+    QVariant data(const QModelIndex&, int) const override { return QVariant(); }
+    int rowCount(const QModelIndex&) const override { return m_padModels.count(); }
+
+    Q_INVOKABLE PercussionPanelPadModel* padModelForIndex(int index) const;
```
Comment (RomanPudashkin) [inline src/notation/view/percussionpanel/percussionpanelpadlistmodel.h:43]: Please re-implement data() instead

File: src/notation/view/percussionpanel/percussionpanelpadlistmodel.cpp [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+{
+    m_padModels = defaultItems();
+}
+
+void PercussionPanelPadListModel::addRow()
+{
+    for (size_t i = 0; i < NUM_COLUMNS; ++i) {
+        m_padModels.append(new PercussionPanelPadModel(QObject::parent()));
+    }
+    emit layoutChanged();
```
Comment (RomanPudashkin) [inline src/notation/view/percussionpanel/percussionpanelpadlistmodel.cpp:36]: We need to use something that allows us to update the view more locally (for example, begin/endRemoveRows)

File: src/notation/view/percussionpanel/percussionpanelpadlistmodel.cpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+void PercussionPanelPadListModel::resetLayout()
+{
+    beginResetModel();
+    m_padModels = defaultItems();
+    endResetModel();
+}
+
+QList<PercussionPanelPadModel*> PercussionPanelPadListModel::defaultItems() const
+{
+    QMap<size_t, QString> instruments;
```
Comment (RomanPudashkin) [inline src/notation/view/percussionpanel/percussionpanelpadlistmodel.cpp:78]: dummy data?

File: src/notation/view/percussionpanel/percussionpanelpadlistmodel.h [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#ifndef MU_NOTATION_PERCUSSIONPANELPADLISTMODEL_H
```
Comment (RomanPudashkin) [inline src/notation/view/percussionpanel/percussionpanelpadlistmodel.h:23]: You can use #pragma once (same for other headers)

File: src/notation/view/percussionpanel/percussionpanelpadlistmodel.cpp (hunk 4 above)
Comment (mathesoncalum) [reply src/notation/view/percussionpanel/percussionpanelpadlistmodel.cpp:78]: Yep!

File: src/notation/view/percussionpanel/percussionpanelpadlistmodel.cpp [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        }
+    }
+
+    // Round up to nearest multiple of 8
+    maxIndex = std::ceil(maxIndex / (double)NUM_COLUMNS) * NUM_COLUMNS;
+
+    QList<PercussionPanelPadModel*> padModels;
+
+    for (size_t i = 0; i < maxIndex; ++i) {
+        PercussionPanelPadModel* model = new PercussionPanelPadModel(QObject::parent());
```
Comment (RomanPudashkin) [inline src/notation/view/percussionpanel/percussionpanelpadlistmodel.cpp:103]: new PercussionPanelPadModel **(this)**

File: src/notation/view/percussionpanel/percussionpanelmodel.cpp [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+{
+    if (itemId == INSTRUMENT_NAMES_CODE) {
+        setUseNotationPreview(false);
+    } else if (itemId == NOTATION_PREVIEW_CODE) {
+        setUseNotationPreview(true);
+    } else if (itemId == EDIT_LAYOUT_CODE) {
+        setCurrentPanelMode(PanelMode::Mode::EDIT_LAYOUT);
+    } else if (itemId == RESET_LAYOUT_CODE) {
+        m_padListModel->resetLayout();
+    } else {
```
Comment (RomanPudashkin) [inline src/notation/view/percussionpanel/percussionpanelmodel.cpp:126]: else is unnecessary here

File: src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+            navigation.section: root.navigationSection
+            navigation.order: root.contentNavigationPanelOrderStart
+
+            model: percModel
+
+            panelWidth: root.width
+        }
+    }
+
+    Item {
```
Comment (RomanPudashkin) [inline src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml:76]: Do we really need this Item? It seems to me that you can use StyledGridView directly here, rather than making it nested in another component

File: src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+            property int numRows: Math.floor(model.rowCount() / numColumns) + 1
+
+            anchors.centerIn: parent
+
+            height: parent.height
+            width: cellWidth * numColumns
+
+            clip: false
+
+            cellWidth: 125
```
Comment (RomanPudashkin) [inline src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml:102]: 126 or 128 (we should avoid numbers that are not multiples of 2 or 4)

File: src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+            delegate: Item {
+                id: delegateRoot
+
+                property int row: Math.floor(index / padGrid.numColumns)
+
+                width: padGrid.cellWidth
+                height: padGrid.cellHeight
+
+                Item {
```
Comment (RomanPudashkin) [inline src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml:115]: Looks like this item is redundant too

File: src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml (hunk 9 above)
Comment (mathesoncalum) [reply src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml:102]: I actually messed up here - it should be `100 + padGrid.spacing` (spacing being 20)

File: src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+                    id: padArea
+
+                    width: parent.width - padGrid.spacing
+                    height: parent.height - padGrid.spacing
+
+                    FlatButton {
```
Comment (RomanPudashkin) [inline src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml:121]: These buttons should not be part of the delegate. We should turn them into separate components, and use only PercussionPanelPad as the delegate

File: src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+
+        clip: true
+
+        StyledGridView {
+            id: padGrid
+
+            readonly property int numColumns: model.numColumns
+            readonly property int spacing: cellWidth / 6
+
+            property int numRows: Math.floor(model.rowCount() / numColumns) + 1
```
Comment (RomanPudashkin) [inline src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml:93]: **readonly** property int numRows

File: src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml (hunk 8 above)
Comment (mathesoncalum) [reply src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml:71]: `dummyToolBarArea` was purely for aesthetic reasons to get the lighter background without needing to change the `PercussionPanelToolBar` component. Not important though, have removed it.

`dummyPanelArea` is handy for handling clipping at the moment, but won't be necessary once I've sorted out the [buttons in delegates](https://github.com/musescore/MuseScore/pull/24753#discussion_r1766743677) problem.

File: src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+        RowLayout {
+            id: rowLayout
+
+            // side columns being the "delete row" buttons on the left, and the "add row" button on the right
+            readonly property int sideColumnsWidth: addRowButton.width
+
+            height: padGrid.cellHeight * padGrid.numRows
+            spacing: padGrid.spacing / 2
+
+            ColumnLayout {
```
Comment (RomanPudashkin) [inline src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml:93]: I think we can use normal Column here

File: src/notation/view/percussionpanel/percussionpanelpadlistmodel.cpp [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "percussionpanelpadlistmodel.h"
+
+PercussionPanelPadListModel::PercussionPanelPadListModel(QObject* parent)
+    : QAbstractListModel(parent)
+{
+    m_padModels = defaultItems();
```
Comment (RomanPudashkin) [inline src/notation/view/percussionpanel/percussionpanelpadlistmodel.cpp:28]: It's better to add Q_INVOKABLE load() and call it from Component.onCompleted instead of filling the model inside the constructor

File: src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+                readonly property int spacing: 20
+
+                Layout.alignment: Qt.AlignTop
+                Layout.fillHeight: true
+
+                width: cellWidth * numColumns
+
+                interactive: false
+
+                cellWidth: 100 + spacing
```
Comment (RomanPudashkin) [inline src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml:146]: padGrid.spacing

File: src/notation/qml/MuseScore/NotationScene/internal/PercussionPanelPad.qml [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+        }
+    }
+
+    states: [
+        State {
+            name: "WRITE"
+            when: root.panelMode === PanelMode.WRITE
+            PropertyChanges {
+                target: modeLabel
+                text: "Write mode"
```
Comment (RomanPudashkin) [inline src/notation/qml/MuseScore/NotationScene/internal/PercussionPanelPad.qml:85]: Untranslatable string

File: src/notation/qml/MuseScore/NotationScene/internal/PercussionPanelPad.qml [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+
+    border.color: root.panelMode === PanelMode.EDIT_LAYOUT ? ui.theme.accentColor : "transparent"
+    border.width: 2
+
+    Item {
+        id: contentArea
+
+        anchors.fill: parent
+
+        visible: root.padModel ? !root.padModel.isEmptySlot : false
```
Comment (RomanPudashkin) [inline src/notation/qml/MuseScore/NotationScene/internal/PercussionPanelPad.qml:48]:
```suggestion
        visible: Boolean(root.padModel) && !root.padModel.isEmptySlot
```

File: src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+    // TODO: Will live inside percussion panel until #22050 is implemented
+    PercussionPanelToolBar {
+        id: toolbar
+
+        anchors.top: parent.top
+
+        width: parent.width
+        height: 35
```
Comment (RomanPudashkin) [inline src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml:62]: 36 maybe?

File: src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+                        FlatButton {
+                            id: deleteButton
+
+                            anchors.verticalCenter: parent.verticalCenter
+                            anchors.right: parent.right
+
+                            visible: model.index > 0
+
+                            icon: IconCode.DELETE_TANK
+                            backgroundRadius: width / 2
```
Comment (RomanPudashkin) [inline src/notation/qml/MuseScore/NotationScene/PercussionPanel.qml:123]: deleteButton.width?

File: src/notation/qml/MuseScore/NotationScene/internal/PercussionPanelToolBar.qml [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+
+Item {
+    id: root
+
+    required property var model
+    required property int panelWidth
+
+    property alias navigation: navPanel
+
+    // anchors.fill: parent TODO: un-comment when #22050 is finished
```
Comment (RomanPudashkin) [inline src/notation/qml/MuseScore/NotationScene/internal/PercussionPanelToolBar.qml:37]: I don't think we should uncomment it :) We should always set the anchors at a higher level, not inside a component

File: src/notation/view/percussionpanel/percussionpanelpadlistmodel.cpp [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+    beginResetModel();
+    m_padModels = defaultItems();
+    endResetModel();
+
+    emit numPadsChanged();
+}
+
+QList<PercussionPanelPadModel*> PercussionPanelPadListModel::defaultItems()
+{
+    QMap<size_t, QString> instruments;
```
Comment (RomanPudashkin) [inline src/notation/view/percussionpanel/percussionpanelpadlistmodel.cpp:97]:
```suggestion
    QMap<size_t, QString> dummyInstruments;
```

File: src/notation/view/percussionpanel/percussionpanelpadlistmodel.h [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <QAbstractListModel>
+
+#include "percussionpanelpadmodel.h"
+
+static const int NUM_COLUMNS(8);
```
Comment (RomanPudashkin) [inline src/notation/view/percussionpanel/percussionpanelpadlistmodel.h:29]:
```suggestion
static constexpr int NUM_COLUMNS(8);
```

File: src/notation/view/percussionpanel/percussionpanelpadlistmodel.h [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+    Q_INVOKABLE void deleteRow(int row);
+
+    Q_INVOKABLE void startDrag(int startIndex);
+    Q_INVOKABLE void endDrag(int endIndex);
+    Q_INVOKABLE bool isDragActive() const;
+
+    int numColumns() const { return NUM_COLUMNS; }
+    int numPads() const { return m_padModels.count(); }
+
+    void resetLayout();
```
Comment (RomanPudashkin) [inline src/notation/view/percussionpanel/percussionpanelpadlistmodel.h:55]: Shouldn't it be private instead?

File: src/notation/view/percussionpanel/percussionpanelpadlistmodel.cpp [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+
+PercussionPanelPadListModel::PercussionPanelPadListModel(QObject* parent)
+    : QAbstractListModel(parent)
+{
+    m_padModels = defaultItems();
+}
+
+QVariant PercussionPanelPadListModel::data(const QModelIndex& index, int role) const
+{
+    if (index.row() < 0 || index.row() >= m_padModels.count()) {
```
Comment (RomanPudashkin) [inline src/notation/view/percussionpanel/percussionpanelpadlistmodel.cpp:33]: return QVariant();

File: src/notation/qml/MuseScore/NotationScene/internal/PercussionPanelToolBar.qml [hunk 25]
```diff
@@ -0,0 +1,154 @@
...
+        accentButton: true
+
+        navigation.panel: navPanel
+        navigation.row: 0
+
+        onClicked: {
+            root.model.finishEditing()
+        }
+    }
+
```
Comment (RomanPudashkin) [inline src/notation/qml/MuseScore/NotationScene/internal/PercussionPanelToolBar.qml:109]: Unnecessary line

File: src/notation/qml/MuseScore/NotationScene/internal/PercussionPanelPad.qml (hunk 16 above)
Comment (mathesoncalum) [reply src/notation/qml/MuseScore/NotationScene/internal/PercussionPanelPad.qml:89]: At the moment these strings are purely a placeholder - they reflect the current state of the pad/panel (I've added some comments detailing this).

File: src/notation/view/percussionpanel/percussionpanelpadlistmodel.h (hunk 23 above)
Comment (mathesoncalum) [reply src/notation/view/percussionpanel/percussionpanelpadlistmodel.h:57]: This method is called by PercussionPanelModel - so it needs to be public.

**Review by RomanPudashkin — APPROVED** [review]
